lightning/config/configlist.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"container/list"
+	"context"
+	"sync"
+)
+
+// ConfigList is a goroutine-safe FIFO list of *Config, which supports removal
+// from the middle. The list is not expected to be very long.
+type ConfigList struct {
+	cond      *sync.Cond
+	nextID    uint32
+	taskIDMap map[uint32]*list.Element
+	nodes     list.List
+}
+
+// NewConfigList creates a new ConfigList instance.
+func NewConfigList() *ConfigList {
+	return &ConfigList{
+		cond:      sync.NewCond(new(sync.Mutex)),
+		taskIDMap: make(map[uint32]*list.Element),
+	}
+}
+
+// Push adds a configuration to the end of the list. The field `cfg.TaskID` will
+// be modified to include a unique ID to identify this task.
+func (cl *ConfigList) Push(cfg *Config) {
+	cl.cond.L.Lock()
+	defer cl.cond.L.Unlock()
+	id := cl.nextID
+	cl.nextID++
+	cfg.TaskID = id
+	cl.taskIDMap[id] = cl.nodes.PushBack(cfg)
+	cl.cond.Broadcast()
+}
+
+// Pop removes a configuration from the front of the list. If the list is empty,
+// this method will block until either another goroutines calls Push() or the
+// input context expired.
+//
+// If the context expired, the error field will contain the error from context.
+func (cl *ConfigList) Pop(ctx context.Context) (*Config, error) {
+	res := make(chan *Config)
+
+	go func() {
+		cl.cond.L.Lock()
+		defer cl.cond.L.Unlock()
+		for {
+			if front := cl.nodes.Front(); front != nil {
+				cfg := front.Value.(*Config)
+				delete(cl.taskIDMap, cfg.TaskID)
+				cl.nodes.Remove(front)
+				res <- cfg
+				break
+			}
+			cl.cond.Wait()
+		}
+	}()
+
+	select {
+	case cfg := <-res:
+		return cfg, nil
```

Comment:
We need to close the `channel` before returning.

Reply:
Closing a channel is not necessary https://stackoverflow.com/questions/8593645/is-it-ok-to-leave-a-channel-open